APICURIO_STORAGE_SQL_KIND=postgresql no longer activates the datasource in 3.3.0 — hardcoded quarkus.datasource.*.active defaults defeat DatabaseConfigInterceptor

## Description

**Registry Version**: 3.3.0 (also reproducible with `latest`, built 2026-06-08)
**Persistence type**: sql

Setting `APICURIO_STORAGE_KIND=sql` and `APICURIO_STORAGE_SQL_KIND=postgresql` (per the
[SQL storage installation docs](https://www.apicur.io/registry/docs/apicurio-registry/3.3.x/getting-started/assembly-installing-registry-docker.html#installing-registry-sql-storage_registry))
no longer activates the `postgresql` named datasource. The registry fails at startup with:

```
io.quarkus.arc.InactiveBeanException: Bean is not active: SYNTHETIC bean [class=io.agroal.api.AgroalDataSource, ...]
Reason: Datasource 'postgresql' was deactivated through configuration properties.
To activate the datasource, set configuration property 'quarkus.datasource."postgresql".active' to 'true'
```

This worked in 3.2.x and earlier. It is a regression introduced in 3.3.0.

### Root cause

`DatabaseConfigInterceptor` is supposed to derive `quarkus.datasource.<kind>.active` from
`apicurio.storage.sql.kind`, and only "provides a default when the property has not been
explicitly set by a higher-priority source" (its Javadoc).

However, PR #7679 added hardcoded defaults to `app/src/main/resources/application.properties`
(present in the 3.3.0 tag, lines 243–246, under a comment saying
"Datasource activation (managed by DatabaseConfigInterceptor based on apicurio.storage.sql.kind)"):

```properties
quarkus.datasource.h2.active=true
quarkus.datasource.postgresql.active=false
quarkus.datasource.mysql.active=false
quarkus.datasource.mssql.active=false
```

The interceptor's "already explicitly set?" check is:

```java
ConfigValue existing = context.proceed(name);
if (existing != null && existing.getValue() != null) {
    return existing;
}
```

`context.proceed()` resolves the *entire* remaining config chain — including
`application.properties` — so it cannot distinguish a user override (env var) from the
application's own baked-in defaults. With the defaults now present, `existing` is always
non-null, the interceptor returns `false` for `postgresql.active`, and the
`apicurio.storage.sql.kind`-based derivation is unreachable dead code.

(Prior to 3.3.0 the `quarkus.datasource.*.active` keys were absent from
`application.properties`, so `proceed()` returned null and the interceptor worked.)

Note: PR #8061 extended the interceptor with `quarkus.datasource.<kind>.jdbc.telemetry`
handling and unit tests, but the tests exercise the interceptor in isolation and don't load
the real `application.properties`, so the regression wasn't caught. The telemetry derivation
is presumably also dead for any key defined in `application.properties`.

### Environment

Docker (Compose), image `apicurio/apicurio-registry:3.3.0`, PostgreSQL 17.

### Steps to Reproduce

1. Start PostgreSQL:
   ```bash
   docker network create apicurio-test
   docker run -d --rm --name pg --network apicurio-test \
     -e POSTGRES_USER=apicurio -e POSTGRES_PASSWORD=apicurio -e POSTGRES_DB=apicurio \
     postgres:17-alpine
   ```
2. Start the registry exactly as documented for SQL storage:
   ```bash
   docker run -it --rm --network apicurio-test \
     -e APICURIO_STORAGE_KIND=sql \
     -e APICURIO_STORAGE_SQL_KIND=postgresql \
     -e APICURIO_DATASOURCE_URL=jdbc:postgresql://pg:5432/apicurio \
     -e APICURIO_DATASOURCE_USERNAME=apicurio \
     -e APICURIO_DATASOURCE_PASSWORD=apicurio \
     -p 8080:8080 apicurio/apicurio-registry:3.3.0
   ```
3. Startup fails with `InactiveBeanException` (see Logs). With 3.2.x the same configuration
   starts successfully.

### Expected vs Actual Behaviour

**Expected**: `APICURIO_STORAGE_SQL_KIND=postgresql` activates the `postgresql` named
datasource (as in ≤3.2.x and as the interceptor's comment/Javadoc intends), and the registry
starts using PostgreSQL storage.

**Actual**: the `postgresql` datasource stays `active=false`, the (active) `h2` datasource is
fed the PostgreSQL JDBC URL (`Datasource 'h2': Driver does not support the provided URL:
jdbc:postgresql://...`), and the application fails to start with `InactiveBeanException`.

**Workaround**: explicitly override the active flags via env vars, which legitimately outrank
`application.properties`:

```bash
-e QUARKUS_DATASOURCE_POSTGRESQL_ACTIVE=true \
-e QUARKUS_DATASOURCE_H2_ACTIVE=false
```

With these set, 3.3.0 starts and persists to PostgreSQL correctly.

### Logs

```
WARN  [io.agroal.pool] Datasource 'h2': Driver does not support the provided URL: jdbc:postgresql://pg:5432/apicurio
INFO  [io.apicurio.registry.storage.impl.sql.RegistryDatasourceProducer] Using postgresql SQL storage.
WARN  [io.apicurio.registry.metrics.health.liveness.PersistenceExceptionLivenessCheck] Liveness problem suspected in PersistenceExceptionLivenessCheck because of an exception:
ERROR [io.quarkus.runtime.Application] Failed to start application
java.lang.RuntimeException: Failed to start quarkus
...
Caused by: io.quarkus.arc.InactiveBeanException: Bean is not active: SYNTHETIC bean [class=io.agroal.api.AgroalDataSource, id=7A6oV7auUZt-GbnaYsjIiA0BlZ4]
Reason: Datasource 'postgresql' was deactivated through configuration properties.
To activate the datasource, set configuration property 'quarkus.datasource."postgresql".active' to 'true' and configure datasource 'postgresql'.
Refer to https://quarkus.io/guides/datasource for guidance.
```

--- a/app/src/main/resources/application.properties
+++ b/app/src/main/resources/application.properties
@@ -239,12 +239,6 @@ apicurio.storage.sql.kind=h2
 apicurio.sql.init=true
 apicurio.sql.db-schema=*
 
-## Datasource activation (managed by DatabaseConfigInterceptor based on apicurio.storage.sql.kind)
-quarkus.datasource.h2.active=true
-quarkus.datasource.postgresql.active=false
-quarkus.datasource.mysql.active=false
-quarkus.datasource.mssql.active=false
-
 apicurio.datasource.url=jdbc:h2:mem:db_${quarkus.uuid}
 apicurio.datasource.username=sa
 apicurio.datasource.password=sa

--- a/app/src/test/java/io/apicurio/registry/event/kafkasql/KafkaSqlEventsTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/event/kafkasql/KafkaSqlEventsTestProfile.java
@@ -11,8 +11,6 @@ public class KafkaSqlEventsTestProfile implements QuarkusTestProfile {
     @Override
     public Map<String, String> getConfigOverrides() {
         return Map.of("apicurio.storage.kind", "kafkasql",
-                "quarkus.datasource.h2.active", "true",
-                "quarkus.datasource.postgresql.active", "false",
                 "apicurio.rest.deletion.artifact.enabled", "true",
                 "apicurio.rest.deletion.artifact-version.enabled", "true",
                 "apicurio.rest.deletion.group.enabled", "true");

--- a/app/src/test/java/io/apicurio/registry/event/sql/EventsTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/event/sql/EventsTestProfile.java
@@ -11,8 +11,6 @@ public class EventsTestProfile implements QuarkusTestProfile {
     @Override
     public Map<String, String> getConfigOverrides() {
         return Map.of("apicurio.storage.sql.kind", "postgresql",
-                "quarkus.datasource.h2.active", "false",
-                "quarkus.datasource.postgresql.active", "true",
                 "apicurio.rest.deletion.artifact.enabled", "true",
                 "apicurio.rest.deletion.artifact-version.enabled", "true",
                 "apicurio.rest.deletion.group.enabled", "true");

--- a/app/src/test/java/io/apicurio/registry/storage/util/KubernetesOpsTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/storage/util/KubernetesOpsTestProfile.java
@@ -12,9 +12,7 @@ public class KubernetesOpsTestProfile implements QuarkusTestProfile {
     public Map<String, String> getConfigOverrides() {
         return Map.of(
                 "apicurio.storage.sql.kind", "h2",
-                "apicurio.storage.kind", "kubernetesops",
-                "quarkus.datasource.h2.active", "true",
-                "quarkus.datasource.postgresql.active", "false"
+                "apicurio.storage.kind", "kubernetesops"
         );
     }
 

--- a/app/src/test/java/io/apicurio/registry/storage/util/MssqlTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/storage/util/MssqlTestProfile.java
@@ -11,9 +11,7 @@ public class MssqlTestProfile implements QuarkusTestProfile {
 
     @Override
     public Map<String, String> getConfigOverrides() {
-        return Map.of("apicurio.storage.sql.kind", "mssql",
-                "quarkus.datasource.h2.active", "false",
-                "quarkus.datasource.mssql.active", "true");
+        return Map.of("apicurio.storage.sql.kind", "mssql");
     }
 
     @Override

--- a/app/src/test/java/io/apicurio/registry/storage/util/MysqlTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/storage/util/MysqlTestProfile.java
@@ -11,9 +11,7 @@ public class MysqlTestProfile implements QuarkusTestProfile {
 
     @Override
     public Map<String, String> getConfigOverrides() {
-        return Map.of("apicurio.storage.sql.kind", "mysql",
-                "quarkus.datasource.h2.active", "false",
-                "quarkus.datasource.mysql.active", "true");
+        return Map.of("apicurio.storage.sql.kind", "mysql");
     }
 
     @Override

--- a/app/src/test/java/io/apicurio/registry/storage/util/PostgresqlTestProfile.java
+++ b/app/src/test/java/io/apicurio/registry/storage/util/PostgresqlTestProfile.java
@@ -11,9 +11,7 @@ public class PostgresqlTestProfile implements QuarkusTestProfile {
 
     @Override
     public Map<String, String> getConfigOverrides() {
-        return Map.of("apicurio.storage.sql.kind", "postgresql",
-                "quarkus.datasource.h2.active", "false",
-                "quarkus.datasource.postgresql.active", "true");
+        return Map.of("apicurio.storage.sql.kind", "postgresql");
     }
 
     @Override

--- a/integration-tests/src/test/resources/infra/sql/registry-sql-secured.yml
+++ b/integration-tests/src/test/resources/infra/sql/registry-sql-secured.yml
@@ -23,10 +23,6 @@ spec:
                   fieldPath: metadata.namespace
             - name: APICURIO_STORAGE_SQL_KIND
               value: "postgresql"
-            - name: QUARKUS_DATASOURCE_H2_ACTIVE
-              value: "false"
-            - name: QUARKUS_DATASOURCE_POSTGRESQL_ACTIVE
-              value: "true"
             - name: APICURIO_DATASOURCE_URL
               value: jdbc:postgresql://postgresql-service:5432/registry
             - name: APICURIO_DATASOURCE_USERNAME

--- a/integration-tests/src/test/resources/infra/sql/registry-sql.yml
+++ b/integration-tests/src/test/resources/infra/sql/registry-sql.yml
@@ -23,10 +23,6 @@ spec:
                   fieldPath: metadata.namespace
             - name: APICURIO_STORAGE_SQL_KIND
               value: "postgresql"
-            - name: QUARKUS_DATASOURCE_H2_ACTIVE
-              value: "false"
-            - name: QUARKUS_DATASOURCE_POSTGRESQL_ACTIVE
-              value: "true"
             - name: APICURIO_DATASOURCE_URL
               value: jdbc:postgresql://postgresql-service:5432/registry
             - name: APICURIO_DATASOURCE_USERNAME

--- a/operator/controller/src/main/java/io/apicurio/registry/operator/feat/SqlStorage.java
+++ b/operator/controller/src/main/java/io/apicurio/registry/operator/feat/SqlStorage.java
@@ -55,9 +55,6 @@ public class SqlStorage {
                         addEnvVar(env, new EnvVarBuilder().withName(ENV_APICURIO_STORAGE_SQL_KIND)
                                 .withValue(sqlKind).build());
 
-                        // Explicitly activate the appropriate Quarkus named datasource
-                        configureDatasourceActivation(storageType, env);
-
                         // Configure datasource connection details
                         addEnvVar(env, new EnvVarBuilder().withName(ENV_APICURIO_DATASOURCE_URL)
                                 .withValue(url).build());
@@ -117,47 +114,4 @@ public class SqlStorage {
         }
     }
 
-    /**
-     * Configures Quarkus datasource activation flags based on the storage type.
-     * Explicitly activates the correct named datasource and deactivates others.
-     *
-     * @param storageType the storage type from the CR
-     * @param env the environment variables map to populate
-     */
-    private static void configureDatasourceActivation(StorageType storageType, Map<String, EnvVar> env) {
-        // H2 is always inactive for operator-managed deployments
-        addEnvVar(env, new EnvVarBuilder()
-                .withName("QUARKUS_DATASOURCE_H2_ACTIVE")
-                .withValue("false")
-                .build());
-
-        // Activate the appropriate datasource based on storage type
-        if (storageType == StorageType.POSTGRESQL) {
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_POSTGRESQL_ACTIVE")
-                    .withValue("true")
-                    .build());
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_MYSQL_ACTIVE")
-                    .withValue("false")
-                    .build());
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_MSSQL_ACTIVE")
-                    .withValue("false")
-                    .build());
-        } else if (storageType == StorageType.MYSQL) {
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_POSTGRESQL_ACTIVE")
-                    .withValue("false")
-                    .build());
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_MYSQL_ACTIVE")
-                    .withValue("true")
-                    .build());
-            addEnvVar(env, new EnvVarBuilder()
-                    .withName("QUARKUS_DATASOURCE_MSSQL_ACTIVE")
-                    .withValue("false")
-                    .build());
-        }
-    }
 }

--- a/utils/tests/src/main/java/io/apicurio/registry/utils/tests/KafkasqlAuthTestProfile.java
+++ b/utils/tests/src/main/java/io/apicurio/registry/utils/tests/KafkasqlAuthTestProfile.java
@@ -11,8 +11,6 @@ public class KafkasqlAuthTestProfile implements QuarkusTestProfile {
     @Override
     public Map<String, String> getConfigOverrides() {
         return Map.of("apicurio.storage.kind", "kafkasql",
-                "quarkus.datasource.h2.active", "true",
-                "quarkus.datasource.postgresql.active", "false",
                 "apicurio.rest.deletion.group.enabled", "true",
                 "apicurio.rest.deletion.artifact.enabled", "true",
                 "apicurio.rest.deletion.artifact-version.enabled", "true");

--- a/utils/tests/src/main/java/io/apicurio/registry/utils/tests/KafkasqlTestProfile.java
+++ b/utils/tests/src/main/java/io/apicurio/registry/utils/tests/KafkasqlTestProfile.java
@@ -10,9 +10,7 @@ public class KafkasqlTestProfile implements QuarkusTestProfile {
 
     @Override
     public Map<String, String> getConfigOverrides() {
-        return Map.of("apicurio.storage.kind", "kafkasql",
-                "quarkus.datasource.h2.active", "true",
-                "quarkus.datasource.postgresql.active", "false");
+        return Map.of("apicurio.storage.kind", "kafkasql");
     }
 
     @Override

